Organize message three-dot menu and change topic editing to be done from modal only.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit is to reorganize the three-dot menu as per #23076.
- Second commit is a simple refactor to not pass unused variable.
- Third commit is to allow only content editing from message edit UI and not topic or stream editing.
- Fourth commit is to change topic editing UI to be a modal.
- Fifth commit is to show "Move message" icon instead of "View source" icon if only moving message is allowed.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures**
![message-popover](https://user-images.githubusercontent.com/35494118/193838972-2047ada9-cf5d-4626-a7d9-ac99fc139302.gif)



**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

alya commented:
While we're here, could we rename "Un-collapse" -> "Expand message"? Thanks!

alya commented:
I think we need to finish #23112 together with these changes?

alya commented:
Also, somehow I'm not seeing the question about which messages to move. I'm not sure what I'm doing differently, since I do see it in the PR description.

sahil839 commented:
I planned to post some comments/questions but was waiting for CI to pass but got occupied with some other work. It failed again, will fix it.

Regarding the dropdown for which messages will be moved, it should be visible once you change stream or topic like we do in message edit UI currently. I was going to ask this whether we want to show it always.

alya commented:
>  I was going to ask this whether we want to show it always.

Yes, we should show it always! I don't know if there's something left over from when we showed it conditionally, but we should be always showing it already after #19196 / #22289?

sahil839 commented on static/templates/actions_popover_content.hbs:
```diff
@@ -27,21 +28,27 @@
     {{/if}}
 
-    {{#if view_source_menu_item}}
+    {{#if (or editability_menu_item move_message_menu_item should_display_delete_option)}}
```
Cannot find any better way to do this. We can also compute it in javascript but yeah this seems atleast better than that.

sahil839 commented on static/js/stream_popover.js:
```diff
@@ -436,24 +440,47 @@ function build_move_topic_to_stream_popover(current_stream_id, topic_name) {
-                    new_topic_name = undefined;
-                }
-
                 if (old_topic_name && select_stream_id) {
```
I have not added these conditions above in the `message !== undefined` block, as I am not sure whether it is required.
`old_topic_name` should always be non-null string. And for `select_stream_id` also, there is no reset button for this widget.

timabbott replied:
I'd look at the history for why this code is there; one possibility is that `new_topic-name` is what was meant, since that's the parameter used in the block inside this conditional? I'm not sure I understand.

sahil839 commented on static/js/ui_init.js:
```diff
@@ -142,10 +142,15 @@ function message_hover($message_row) {
+        editability === message_edit.editability_types.TOPIC_ONLY || is_stream_editable;
     const args = {
-        is_editable: is_message_editable && !message.status_message,
+        is_content_editable: is_content_editable && !message.status_message,
```
Not sure why we check `status_message` here, we allow editing `/me` messages currently.

sahil839 commented:
Posted some comments above and have completed all parts of #23112. Also done following changes as requested above-
- The dropdown for what messages to move is shown always.
- Added commit to change `Un-collapse` to `Expand message`.
- Also we need to add backend changes for sending the notification messages when only topic is changed and not streams. I guess there is already an open PR for it.

sahil839 commented:
> Also we need to add backend changes for sending the notification messages when only topic is changed and not streams. I guess there is already an open PR for it.

#22312

alya commented:
> > Also we need to add backend changes for sending the notification messages when only topic is changed and not streams. I guess there is already an open PR for it.
> 
> #22312

I tagged the issue for this (#21712) as a release goal.

alya commented:
Could we give more space to the "what messages to move" dropdown? It would be good if all the options could fit horizontally as long as there's room for that in the modal. Separately, the `g` looks a bit cut off vertically?

<img width="360" alt="Screen Shot 2022-10-06 at 11 41 46 AM" src="https://user-images.githubusercontent.com/2090066/194393520-22e67cd9-8fe8-4e48-b056-8e6671e92b4f.png">

alya commented:
Could we add explanatory tooltips on-hover over a greyed out stream or topic in the "Move messages" modal? Something like:

- "You do not have permission to move messages to a different stream."
- "You do not have permission to change the topic name."

I also noticed that the stream field (but not the topic field) has hover behavior even when it's disabled. I think we don't need hover behavior for a disabled stream selector?

sahil839 commented:
Updated the PR.
- Increased the width of "what messages to move?" dropdown. Its width will be to fit the longest option till it remains inside the modal.
![Screenshot from 2022-10-07 22-30-29](https://user-images.githubusercontent.com/35494118/194611712-089d3b62-0c68-4cb5-9bfb-6912c40f3e14.png)
- Fixed the stream button to have no hover behavior when it is disabled.
- Currently we do not show the `Move topic` option in left sidebar when changing stream is not allowed, including the case when topic editing is allowed. This is because we do may not have the first message of the topic loaded in the client to tell whether topic editing is allowed or not. For the same reason we cannot disable the topic input because we cannot check without first making a request to server to get the first message id. But both these problems will be solved as we would be making topic editing permissions independent of message. In the meantime, I have added a commit to instead keep the modal open and show the error inside the modal itself. This behavior should be fine for other type of errors as well.
![Screenshot from 2022-10-07 22-29-57](https://user-images.githubusercontent.com/35494118/194613245-820eb2ad-9206-4ece-9474-f8aa3d40a049.png)

Regarding other points -
- I can't reproduce the letter `g` being cut off from bottom. This dropdown has the same height as other dropdowns in the app.
- For showing tooltips on disable elements we need to add a container div element, and tippy documentation does not recommend to do so because of accessibility concerns.

alya commented:
> Currently we do not show the Move topic option in left sidebar when changing stream is not allowed, including the case when topic editing is allowed. This is because we do may not have the first message of the topic loaded in the client to tell whether topic editing is allowed or not. For the same reason we cannot disable the topic input because we cannot check without first making a request to server to get the first message id. But both these problems will be solved as we would be making topic editing permissions independent of message. In the meantime, I have added a commit to instead keep the modal open and show the error inside the modal itself. This behavior should be fine for other type of errors as well.

That looks like a nice improvement!

Is there a reason we show the "Move topic" menu option at all when message editing is turned off? I think you can't do anything in that menu when message editing is not allowed? Anyway, not worth spending more than a few minutes on this since, as you said, we're revamping that whole permissions system in any case.

alya commented:
> For showing tooltips on disable elements we need to add a container div element, and tippy documentation does not recommend to do so because of accessibility concerns.

Got it. I guess it's OK to skip the tooltips.

alya commented:
@timabbott I think this is ready for your review!

Note that we need to integrate #21712 / #22312 for this set of changes not to feel buggy, where you select a "Notify" option, but no notification is sent.

sahil839 commented on static/js/stream_popover.js:
```diff
@@ -414,46 +426,78 @@ function build_move_topic_to_stream_popover(e, current_stream_id, topic_name) {
+                // important for cases when changing stream is
+                // not allowed.
+                select_stream_id = undefined;
+            }
```
We can move this outside the `message !== undefined` if we can be sure that the [below condition is not necessary](https://github.com/zulip/zulip/pull/23139/files#r989309707).

sahil839 commented on static/js/stream_popover.js:
```diff
@@ -414,46 +426,78 @@ function build_move_topic_to_stream_popover(e, current_stream_id, topic_name) {
+            // We use `undefined` to tell the server that
+            // there has been no change in the topic name.
+            new_topic_name = undefined;
+        }
```
This is somewhat confusing, but `serializeArray` in `get_params_from_form` returns undefined for disabled elements even if the input is not empty and is filled with the old topic itself. We can set `new_topic_name` to `old_topic_name` if `new_topic_name` is undefined but then also it will get changed to `undefined` again due to both being same.

timabbott replied:
This code path is pretty weird and probably worth rethinking later, but this is OK for now.

sahil839 commented on static/templates/edit_content_button.hbs:
```diff
@@ -1,5 +1,7 @@
+<i class="fa fa-arrows move_message_button edit_message_button" role="button" tabindex="0" aria-label="{{t 'Move message' }} (e)" data-tippy-content="{{#tr}}Move message{{/tr}}"></i>
 {{else}}
-<i class="fa fa-file-code-o edit_content_button" role="button" tabindex="0" aria-label="{{t 'View message source' }} (e)" data-tippy-content="{{#tr}}View message source{{/tr}} (e)" data-message-id="{{msg_id}}"></i>
+<i class="fa fa-file-code-o view_source_button edit_message_button" role="button" tabindex="0" aria-label="{{t 'View message source' }} (e)" data-tippy-content="{{#tr}}View message source{{/tr}} (e)" data-message-id="{{msg_id}}"></i>
```
We want to use a common class here for handling the tippy tooltips code, so I have just use `edit_message_button` for view source icon as well, since otherwise something like `edit_or_view_source_message...` might be too long. Can change it if we want to avoid any unnecessary confusion.

timabbott replied:
Yeah, I think that's probably fine.

timabbott commented on templates/zerver/help/keyboard-shortcuts.md:
```diff
@@ -143,7 +143,7 @@ below, and add more to your repertoire as needed.
 * **View image**: <kbd>V</kbd>
 
-* **Edit message**: <kbd>E</kbd>
+* **Edit message or view message source**: <kbd>E</kbd>
```
@alya we might need more explanation or link here; I'm not sure users reading this will understand what "view message source" means here.

timabbott commented:
I did a reasonably thorough read on this and didn't find any significant issues other than wanting to include #22312; let me see if I can get both this and that on chat.zulip.org.

sahil839 commented:
I can rebase and fix the tests in #22312 if it helps. And I guess we can update the frontend changes too which will not be required once this PR is merged.

timabbott commented:
Yeah that'd be very helpful!

sahil839 commented:
> Currently we do not show the Move topic option in left sidebar when changing stream is not allowed, including the case when topic editing is allowed. This is because we do may not have the first message of the topic loaded in the client to tell whether topic editing is allowed or not. For the same reason we cannot disable the topic input because we cannot check without first making a request to server to get the first message id. But both these problems will be solved as we would be making topic editing permissions independent of message. In the meantime, I have added a commit to instead keep the modal open and show the error inside the modal itself. This behavior should be fine for other type of errors as well.

While working on further parts of #21739, I realised the topic editing permission would not be independent of message since we would still need the first message of topic to check whether the time limit has passed or not. And as per #21739 the time limit setting would apply for moving messages across streams also, so we would face problem in deciding when to show the "Move topic" option in left sidebar popover and also for disabling the inputs in modal.

timabbott commented:
I rebased this on top of #22312 after merging it, but in manual testing, it seems like we're not actually sending the "move topic" notifications when moving messages within a stream, and I'm not sure why. 

I guess I'll test-deploy this anyway since I'd like to get folks some exposure to the new menu format.

sahil839 commented:
I tested and I can see the notifications being sent.

timabbott commented:
It's working for me now here and also on chat.zulip.org; I'm not sure what my problem was locally before. But since I couldn't figure out a mechanism for it not working before anyway, probably it was something like the server not having properly restarted.

Merged, thanks @sahil839!
